d3a_api_client/redis_client_base.py

```diff
@@ -2,57 +2,73 @@
 from redis import StrictRedis
+from slugify import slugify
 
 from d3a_api_client import APIClientInterface
 from d3a_api_client.constants import MAX_WORKER_THREADS
+from d3a_api_client.utils import execute_function_util
 
 
 class RedisAPIException(Exception):
     pass
 
 
-def registered_connection(f):
-    @wraps(f)
-    def wrapped(self, *args, **kwargs):
-        if not self.is_active:
-            raise RedisAPIException(f'Registration has not completed yet.')
-        return f(self, *args, **kwargs)
-
-    return wrapped
-
-
-class RedisClient(APIClientInterface):
+class RedisClientBase(APIClientInterface):
     def __init__(self, area_id, autoregister=True, redis_url='redis://localhost:6379',
                  pubsub_thread=None):
         super().__init__(area_id, autoregister, redis_url)
+        self.area_uuid = None
+        self.area_slug = slugify(area_id, to_lower=True)
         self.redis_db = StrictRedis.from_url(redis_url)
         self.pubsub = self.redis_db.pubsub() if pubsub_thread is None else pubsub_thread
         self.area_id = area_id
         self.device_uuid = None
         self.is_active = False
         self._blocking_command_responses = {}
+        self._transaction_id_buffer = []
+        self._subscribed_aggregator_response_cb = None
         self._subscribe_to_response_channels(pubsub_thread)
         self.executor = ThreadPoolExecutor(max_workers=MAX_WORKER_THREADS)
         if autoregister:
             self.register(is_blocking=True)
 
+    @property
+    def _channel_prefix(self):
+        return f"{self.area_slug}"
+
     @property
     def channel_subs(self):
         return {
-            f"{self.area_id}/response/register_participant": self._on_register,
-            f"{self.area_id}/response/unregister_participant": self._on_unregister,
+            f'{self._channel_prefix}/response/register_participant': self._on_register,
+            f'{self._channel_prefix}/response/unregister_participant': self._on_unregister,
+            f'{self._channel_prefix}/*': self._on_event_or_response
         }
 
     def _subscribe_to_response_channels(self, pubsub_thread=None):
         self.pubsub.psubscribe(**self.channel_subs)
+        if b'aggregator_response' in self.pubsub.patterns:
+            self._subscribed_aggregator_response_cb = self.pubsub.patterns[b'aggregator_response']
         if pubsub_thread is None:
             self.pubsub.run_in_thread(daemon=True)
 
-    def register(self, is_blocking=False):
+    def _check_buffer_message_matching_command_and_id(self, message):
+        if key_in_dict_and_not_none(message, "transaction_id"):
+            transaction_id = message["transaction_id"]
+            if not any(command in ["register", "unregister"] and "transaction_id" in data and
+                       data["transaction_id"] == transaction_id
+                       for command, data in self._blocking_command_responses.items()):
```

Comment:
Nit: this would probably deserve a separate method, to explain better what we're doing in this loop.

Reply:
I wanted to do it, but then I searched for a name of  the function. It would be the same as the function here `_check_buffer_message_matching_command_and_id` . So this is the main part of the functionality. So I vote for leaving it as is.

Reply:
Yeah, this is actually tricky. Maybe we could improve the readability of the method by turning the negative statement into a positive one. So, rather than:
```python
if not any(...):
    raise RedisAPIException(...)
```
we could have:
```python
if all(...):
    return True
else:
    raise RedisAPIException(...)
```
This is sometimes referred as the "Avoid negative conditionals" rule, and I think it helps readability.
Anyway, this is minor so feel free to leave it as it is if you want.

Reply:
@fievelk `not any() != all()` 
We only want to check whether the condition is true for at least one of the responses. Not for all. So i dont see room for improvement here.

Reply:
I think my example should work though. It's not just replacing `not any()` with `all()`, it's also moving the exception into the `else` statement.
So we say:
> "if all the messages match the condition, then return True. Otherwise, raise an exception"

The version with `not any()` is saying:
> "if there is not even one message that matches the condition, then we raise an exception."

I hope I'm not getting confused by all the negations :P

Reply:
Sorry, I was getting confused earlier. In this case, I would just use
```python
if any(...):
    return True
else:
    raise RedisAPIException(...)
```
to avoid negative conditionals. Again, it doesn't need to be changed here but it might be useful in the future. I find these conditionals clearer than the negative ones.